guide/qip/qip-simulator.rst

```diff
@@ -0,0 +1,318 @@
+.. QuTiP
+   Copyright (C) 2011-2012, Paul D. Nation & Robert J. Johansson
+
+.. _qip_simulator:
+
+****************
+Exact Simulation
+****************
+
+Let's start off by defining a simple circuit which we use to demonstrate a few
+examples of circuit evolution. First, we read in the circuit from a **.qasm** file using
+the :func:`qutip.qip.qasm.read_qasm()` function.
+
+.. testcode::
+
+  from qutip.qip.qasm import read_qasm
+  from qutip import tensor, basis
+
+  qc = read_qasm("w-state.qasm")
+
+  qc.png
+
+.. image:: /gallery/qutip_examples/qip/images/w_state.png
+
+This circuit prepares the W-state :math:`\newcommand{\ket}[1]{\left|{#1}\right\rangle} \frac{\ket{001} + \ket{010} + \ket{100}}{\sqrt{3}}`.
+The simplest way to carry out state evolution through a quantum circuit is
+providing a input state to the :func:`qutip.qip.QubitCircuit.run()`
+function.
+
+.. testcode::
+
+  zero_state = tensor(basis(2, 0), basis(2, 0), basis(2, 0))
+  result = qc.run(state=zero_state)
+  wstate = result.get_states()[0]
+
+  print(wstate)
+
+**Output**:
+
+.. testoutput::
+  :options: +NORMALIZE_WHITESPACE
+
+  Quantum object: dims = [[2, 2, 2], [1, 1, 1]], shape = (8, 1), type = ket
+  Qobj data =
+  [[0.+0.j        ]
+   [0.+0.57734961j]
+   [0.+0.57734961j]
+   [0.+0.j        ]
+   [0.+0.57735159j]
+   [0.+0.j        ]
+   [0.+0.j        ]
+   [0.+0.j        ]]
+
+
+As expected, the state returned is indeed the required W-state.
+The function returns a :class:`qutip.qip.Result` object which contains
+the output state which can be obtained using the :func:`qutip.qip.Result.get_states()`
+function (Infact, an array of output states is returned).
+
```

Comment:
If we know that the result is only one Qobj, like from `run`, we can return this Qobj directly without making it in a list. Should be enough to just remove the list when the simulator is called by `run`.

This makes sure that the default behaviour, a circuit without measurement or use only `run`,  the same as it used to be. A single ket `Qobj` as an output.

Reply:
I have changed this to return `Qobj` but only when `run` is called from a circuit directly but returning `CircuitResult` when called from Simulator. It would still be simple enough for beginning users while giving more advanced users opportunity to use as needed.